Host/libeCos/src/net/tcpip/v2_0/doc/select.man
NAME
       select,  FD_CLR,  FD_ISSET,  FD_SET, FD_ZERO - synchronous
       I/O multiplexing

SYNOPSIS
       #include <network.h>

       int  select(int  n,  fd_set  *readfds,  fd_set  *writefds,
       fd_set *exceptfds, struct timeval *timeout);

       FD_CLR(int fd, fd_set *set);
       FD_ISSET(int fd, fd_set *set);
       FD_SET(int fd, fd_set *set);
       FD_ZERO(fd_set *set);

DESCRIPTION
       select  waits  for  a number of file descriptors to change
       status.

       Three independent sets of descriptors are watched.   Those
       listed  in  readfds  will  be watched to see if characters
       become available for reading, those in  writefds  will  be
       watched  to  see if it is ok to immediately write on them,
       and those in exceptfds will be watched for exceptions.  On
       exit,  the  sets  are  modified in place to indicate which
       descriptors actually changed status.

       Four macros are provided to manipulate the sets.   FD_ZERO
       will clear a set.  FD_SET and FD_CLR add or remove a given
       descriptor from  a  set.   FD_ISSET  tests  to  see  if  a
       descriptor is part of the set; this is useful after select
       returns.

       n is the highest-numbered descriptor in any of  the  three
       sets, plus 1.

       timeout  is  an  upper bound on the amount of time elapsed
       before select returns. It may be zero, causing  select  to
       return  immediately.  If  timeout  is  NULL  (no timeout),
       select can block indefinitely.

RETURN VALUE
       On success, select returns the number of descriptors  con-
       tained  in  the  descriptor sets, which may be zero if the
       timeout expires before anything interesting  happens.   On
       error, -1 is returned, and errno is set appropriately; the
       sets and timeout become undefined, so do not rely on their
       contents after an error.

ERRORS
       EBADF   An invalid file descriptor was given in one of the
               sets.

       EINTR   A non blocked signal was caught.

       EINVAL  n is negative.

       ENOMEM  select was unable to allocate memory for  internal
               tables.

NOTES
       Some  code calls select with all three sets empty, n zero,
       and a non-null timeout as a fairly portable way  to  sleep
       with subsecond precision.

EXAMPLE
       #include <stdio.h>
       #include <sys/time.h>
       #include <sys/types.h>
       #include <unistd.h>

       int
       main(void)
       {
           fd_set rfds;
           struct timeval tv;
           int retval;

           /* Watch stdin (fd 0) to see when it has input. */
           FD_ZERO(&rfds);
           FD_SET(0, &rfds);
           /* Wait up to five seconds. */
           tv.tv_sec = 5;
           tv.tv_usec = 0;

           retval = select(1, &rfds, NULL, NULL, &tv);
           /* Don't rely on the value of tv now! */

           if (retval)
               printf("Data is available now.\n");
               /* FD_ISSET(0, &rfds) will be true. */
           else
               printf("No data within five seconds.\n");

           exit(0);
       }

       Generally  portable  to/from  non-BSD  systems  supporting
       clones  of  the BSD socket layer (including System V vari-
       ants).  However, note that the System V variant  typically
       sets the timeout variable before exit, but the BSD variant
       does not.


ChangeLog.txt
/***** DONT REMOVE OR MODIFY THE FOLLOWING 4 LINES! *****/
  "$Revision: 3705 $",
  "$Id: ChangeLog.txt 3705 2010-03-10 01:53:10Z xhchen $",
  "$Build: "__DATE__" "__TIME__ " $"
/********************************************************/

/*
2010 03 10
    1. Add project HelloWorld for libeCos and open libeCos's source.

2010 01 12
    1. Update new HTML.
    2. Fixed the problem of RTSP buffer overflow.
    
2009 12 24
    1. Reset IP trigger event after DHCP renewed.

2009 11 24
    1. Use random delay time before reconnect PPP server.

2009 11 12
    1. Remove USB debug message.
    2. PPP can reconnect on error.

2009 10 04
    1. Fixed VLC/QuickTime can not play RTSP for long time.

2009 10 23
    1. Fixed that cyg_fd_free may access unknown area.
    2. LED use its own lock instead of period task's lock.

2009 10 16
    1. Can output debug message by VCOM.
    2. SetPPP and GetPPP by VCOM.

2009 10 09
    1. Use new ActiveX version 1.0.0.9.

2009 09 17
    1. Can support up to 10 PPP servers.
    2. Win7 can display MP4 and use VCOM. (VCOM still has problems).
    3. More stable ActiveX.
    4. Add timeout for RTSP TCP connection.

2009 08 14
    1. IPCamConfigure can getback password.

2009 08 12
    1. PPP: Fixed the bug of block on fd close.
    2. PPP: Fixed the bug of block on cyg_io_write/splany.

2009 08 04
    1. Add SetPPP.cgi/GetPPP.cgi.
    2. Fixed ActiveX version error.

2009 07 02
    1. Make nupdate workable on XP.

2009 06 05
    1. MJPEG sometimes slow, fixed.

2009 04 27
    1. Make console usb tool and verify script compatible with Rovio 
       v4.02.
    2. Fixed the bug about add/del user and sprintf.

2009 04 24
    1. Fixed problem ActiveX blocks. (HttpSendRequestEx still blocks).
    2. Add the script to change ActiveX version and clsid.

2009 04 21
    1. Apply CP's update for UART test.

2009 04 20
    1. Break USB writing on USB error.
    2. USB console tools:
       a. Fixed the error in percent value of updating firmware.
       b. Return value fo initSerialConnect() is not correct.
    3. USBWrite will always break out after USB cable disconnect.
    4. Adjust priority of wlan main thread to 10.


2009 04 17
    1. Fiex RTSP server's bug that server will crash when quicktime teardown normally.
    2. Reduce delay time to audio test.
    3. Merged from W99803 branch:
       a. Fixed memory leak in sending mail.

2009 04 15
    1. Disable auth check for USB console tool.
    2. Update verify tool:
       a. stdout/stderr was changed since that wowwee flash tool can
          not catch pipe data from stderr.
       b. disable test for suspend.
       c. support new usb interface.

2009 04 08
    1. Use new parameter "Description" for SetVNet.cgi/GetVNet.cgi
    2. Fix a pointer error in bpserver.c

2009 04 03
    1. Fixed sscanf problemm in USB console tool.
    2. Add delay time in SetVNet.cgi

2009 03 24
    1. Add channel number settings for SetVNet.cgi.

2009 03 20
    1. Changed the ActiveX version to 1.0.0.8 and clsid to
       {115B1886-2AE0-4259-9FE4-E32A5DEE5454}

2009 03 19
    1. Modify "user privilege" fields in USB tools, so that it's
       compitable with old USB tools. 
    2. Merge RTSP drop_conn patch from W99803's branch.
    3. Do few things when rebooting.

2009 03 18
    1. Fixed Activex close crash bug.
    2. fixed activex wrong url bug.
    3. updated wbstreamplayer to support mp4 width > 640 height>48.

2009 03 16
    1. Do not disable/enable GPIO13 while accessing SPI flash,
       which may fix the problem of flash corrupted.
    2. Add -f output_file for USB command line tool.

2009 03 13
    1.updated activex for MSI filter.

2009 03 06
    1. Synchronzie with W99803's branch.

2009 03 05
    1. Add username/password for usb console tool.

2009 03 04
    1. Add new command "usb netstate", which reports:
       WIFI state:
           NO_INIT
           TRYING
           OK
       IP state:
           NO_INIT
           DHCP_TRYING
           DHCP_FAILED
           DHCP_OK
           STATIC_IP_TRYING
           STATIC_IP_FAILED
           STATIC_IP_OK
    2. "usb user" reports:
           privilege:username
    3. If a key is used, "usb wlan" reports the key with "*",
       otherwise report the blank string.
    4. Remove password for user delete.

2009 02 27
    1. Set mp4 max width to 1600 ,max height to 1200

2009 02 25
    1. update Activex for MSI facial recognition function.
2009 02 24
    1. Optimize getting mp4/jpeg and audio. 

2009 02 23
    1. Synchronzie with W99803's branch.

2009 02 20
    1. Adjust stdout/stderr in console usb tool.

2009 02 19
    1. Update console usb tool, added scan_wlan, auth enable/disable,
       updated command wlan.

2009 02 18
    1. Use large packet size for H.263.
       ( Bug not fixed for H.263 packetize. )
    2. Add current IP information to GetIP.cgi.

2009 02 13
    1. Clear error state after MCU is restored.
    2. Use timestamp on capture point for audio and video.

2009 02 11
    1. updated input-oss to check frame heap before getting audio.

2009 02 09
    1. Import source of updatepmp and vcom.
    2. Fixed the problem about setting WLAN static IP may
       cause long time waiting.

2009 02 03
    1. update wbstreamplayer for rtsp authentication.
    2. Add save user/pass for ActiveX.

2009 01 22
    1. Add authorization dialog for ActiveX.
       Have not applied JF's update for WBStreamPlayer.

2009 01 21
    1. Add basic auth for RTSP.

2009 01 20
    1. Integrate RTSP authorization with HTTP.
       Enable authorization for RTSP over UDP & TCP.

2009 01 15
    1. Test RTSP over HTTP, OK.
    2. Disable LED low battery on charge. 

2009 01 12
    1. Enlarge VCOM buffer size to 4K.

2009 01 08
    1. Fixed LED error.
    2. Update to libNS 5.13.


Host/libeCos/src/io/fileio/v2_0/src/select.cxx
//==========================================================================
//
//      select.cxx
//
//      Fileio select() support
//
//==========================================================================
//####ECOSGPLCOPYRIGHTBEGIN####
// -------------------------------------------
// This file is part of eCos, the Embedded Configurable Operating System.
// Copyright (C) 1998, 1999, 2000, 2001, 2002 Red Hat, Inc.
// Copyright (C) 2002 Nick Garnett
//
// eCos is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 2 or (at your option) any later version.
//
// eCos is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
//
// You should have received a copy of the GNU General Public License along
// with eCos; if not, write to the Free Software Foundation, Inc.,
// 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
//
// As a special exception, if other files instantiate templates or use macros
// or inline functions from this file, or you compile this file and link it
// with other works to produce a work based on this file, this file does not
// by itself cause the resulting work to be covered by the GNU General Public
// License. However the source code for this file must still be made available
// in accordance with section (3) of the GNU General Public License.
//
// This exception does not invalidate any other reasons why a work based on
// this file might be covered by the GNU General Public License.
//
// Alternative licenses for eCos may be arranged by contacting Red Hat, Inc.
// at http://sources.redhat.com/ecos/ecos-license/
// -------------------------------------------
//####ECOSGPLCOPYRIGHTEND####
//==========================================================================
//#####DESCRIPTIONBEGIN####
//
// Author(s):           nickg
// Contributors:        nickg
// Date:                2000-05-25
// Purpose:             Fileio select() support
// Description:         Support for select().
//                      
//              
//              
//
//####DESCRIPTIONEND####
//
//==========================================================================

#include "pkgconf/hal.h"
#include "pkgconf/kernel.h"
#include "pkgconf/io_fileio.h"

#include "cyg/kernel/ktypes.h"         // base kernel types
#include "cyg/infra/cyg_trac.h"        // tracing macros
#include "cyg/infra/cyg_ass.h"         // assertion macros

#include "stdarg.h"                    // for fcntl()

#include "fio.h"                       // Private header

#include "sys/select.h"                // select header

#include "cyg/kernel/sched.hxx"        // scheduler definitions
#include "cyg/kernel/thread.hxx"       // thread definitions
#include "cyg/kernel/mutex.hxx"        // mutex definitions
#include "cyg/kernel/clock.hxx"        // clock definitions

#include "cyg/kernel/sched.inl"
#include "cyg/kernel/thread.inl"
#include "cyg/kernel/clock.inl"

//==========================================================================
// File object locking

#define LOCK_FILE( fp ) cyg_file_lock( fp )

#define UNLOCK_FILE( fp ) cyg_file_unlock( fp )

//==========================================================================
// Local variables

// Mutex for serializing select processing. This essntially controls
// access to the contents of the selinfo structures embedded in the
// client system data structures.
static Cyg_Mutex select_mutex CYGBLD_ATTRIB_INIT_PRI(CYG_INIT_IO_FS);

// Condition variable where any thread that is waiting for a select to
// fire is suspended. Note that select is not intended to be a real time
// operation. Whenever any selectable event occurs, all selecting threads
// will be resumed. They must then rescan their selectees and resuspend if
// necessary.
static Cyg_Condition_Variable selwait( select_mutex ) CYGBLD_ATTRIB_INIT_PRI(CYG_INIT_IO_FS);

static volatile cyg_uint32 selwake_count = 0;

//==========================================================================
// Timeval to ticks conversion support

// Converters from sec and us to ticks
static struct Cyg_Clock::converter us_converter, sec_converter;

static cyg_bool converters_initialized = false;

externC cyg_tick_count cyg_timeval_to_ticks( const struct timeval *tv )
{
    if( !converters_initialized )
    {
        // Create the converters we need.
        Cyg_Clock::real_time_clock->get_other_to_clock_converter( 1000, &us_converter );
        Cyg_Clock::real_time_clock->get_other_to_clock_converter( 1000000000, &sec_converter );

        converters_initialized = true;
    }
    
    // Short circuit zero timeval
    if( tv->tv_sec == 0 && tv->tv_usec == 0 )
    {
        return 0;
    }
        
    // Convert the seconds field to ticks.
    cyg_tick_count ticks = Cyg_Clock::convert( tv->tv_sec, &sec_converter );

    // Convert the nanoseconds. This will round down to nearest whole tick.
    ticks += Cyg_Clock::convert( (cyg_tick_count)tv->tv_usec, &us_converter );

    return ticks;
}

//==========================================================================
// Select API function

static int
cyg_pselect(int nfd, fd_set *in, fd_set *out, fd_set *ex,
           struct timeval *tv, const sigset_t *mask)
{
    FILEIO_ENTRY();

    int error = ENOERR;
    int fd, mode, num;
    cyg_file *fp;
    fd_set in_res, out_res, ex_res;  // Result sets
    fd_set *selection[3], *result[3];
    cyg_tick_count ticks;
    int mode_type[] = {CYG_FREAD, CYG_FWRITE, 0};
    cyg_uint32 wake_count;
    sigset_t oldmask;

    FD_ZERO(&in_res);
    FD_ZERO(&out_res);
    FD_ZERO(&ex_res);

    // Set up sets
    selection[0] = in;   result[0] = &in_res;
    selection[1] = out;  result[1] = &out_res;
    selection[2] = ex;   result[2] = &ex_res;

    // Compute end time
    if (tv)
        ticks = cyg_timeval_to_ticks( tv );
    else ticks = 0;

    // Lock the mutex
    select_mutex.lock();

    // Scan sets for possible I/O until something found, timeout or error.
    while (!error)
    {
        wake_count = selwake_count;
        
        num = 0;  // Total file descriptors "ready"
        for (mode = 0;  !error && mode < 3;  mode++)
        {
            if (selection[mode]) {
                for (fd = 0;  !error && fd < nfd;  fd++)
                {
                    if (FD_ISSET(fd, selection[mode]))
                    {
                        fp = cyg_fp_get( fd );
                        if( fp == NULL )
                        {
                            error = EBADF;
                            break;
                        }

                        if ((*fp->f_ops->fo_select)(fp, mode_type[mode], 0))
                        {
                            FD_SET(fd, result[mode]);
                            num++;
                        }

                        cyg_fp_free( fp );
                    }
                }
            }
        }
        
        if (num)
        {
            // Found something, update user's sets
            if (in)  FD_COPY( &in_res, in );
            if (out) FD_COPY( &out_res, out );
            if (ex)  FD_COPY( &ex_res, ex );
            select_mutex.unlock();
            CYG_FILEIO_DELIVER_SIGNALS( mask );
            FILEIO_RETURN_VALUE(num);
        }

        Cyg_Scheduler::lock();

        // Switch to the supplied signal mask. This will permit delivery
        // of any signals that might terminate this select operation.
        
        CYG_FILEIO_SIGMASK_SET( mask, &oldmask );
    
        do
        {

            // We need to see if any signals have been posted while we
            // were testing all those files. The handlers will not
            // have run because we have ASRs inhibited but the signal
            // will have been set pending.

            if( CYG_FILEIO_SIGPENDING() )
            {
                // There are pending signals so we need to terminate
                // the select operation and return EINTR. Handlers for
                // the pending signals will be called just before we
                // return.

                error = EINTR;
                break;
            }
            
            if( wake_count == selwake_count )
            {
                // Nothing found, see if we want to wait
                if (tv)
                {
                    // Special case of "poll"
                    if (ticks == 0)
                    {
                        error = EAGAIN;
                        break;
                    }

                    ticks += Cyg_Clock::real_time_clock->current_value();
                
                    if( !selwait.wait( ticks ) )
                    {
                        // A non-standard wakeup, if the current time is equal to
                        // or past the timeout, return zero. Otherwise return
                        // EINTR, since we have been released.

                        if( Cyg_Clock::real_time_clock->current_value() >= ticks )
                        {
                            error = EAGAIN;
                            break;
                        }
                        else error = EINTR;
                    }

                    ticks -= Cyg_Clock::real_time_clock->current_value();
                }
                else
                {
                    // Wait forever (until something happens)
            
                    if( !selwait.wait() )
                        error = EINTR;
                }
            }

        } while(0);

        CYG_FILEIO_SIGMASK_SET( &oldmask, NULL );
        
        Cyg_Scheduler::unlock();
        
    } // while(!error)

    select_mutex.unlock();
 
    // If the error code is EAGAIN, this means that a timeout has
    // happened. We return zero in that case, rather than a proper
    // error code.
    // If the error code is EINTR, then a signal may be pending
    // delivery. Call back into the POSIX package to handle it.
    
    if( error == EAGAIN )
        FILEIO_RETURN_VALUE(0);
    else if( error == EINTR )
        CYG_FILEIO_DELIVER_SIGNALS( mask );

    FILEIO_RETURN(error);
}

// -------------------------------------------------------------------------
// Select API function

__externC int
select(int nfd, fd_set *in, fd_set *out, fd_set *ex, struct timeval *tv)
{
	return cyg_pselect(nfd, in, out, ex, tv, NULL);
}

// -------------------------------------------------------------------------
// Pselect API function
//
// This is derived from the POSIX-200X specification.

#ifdef CYGPKG_POSIX

__externC int
pselect(int nfd, fd_set *in, fd_set *out, fd_set *ex,
	const struct timespec *ts, const sigset_t *sigmask)
{
	struct timeval tv;

	if (ts != NULL)
        {
            tv.tv_sec = ts->tv_sec;
            tv.tv_usec = ts->tv_nsec/1000;
        }

	return cyg_pselect(nfd, in, out, ex, &tv, sigmask);
}

#endif

//==========================================================================
// Select support functions.

// -------------------------------------------------------------------------
// cyg_selinit() is used to initialize a selinfo structure

void cyg_selinit( struct CYG_SELINFO_TAG *sip )
{
    sip->si_info = 0;
    sip->si_thread = 0;
}

// -------------------------------------------------------------------------
// cyg_selrecord() is called when a client device needs to register
// the current thread for selection.

void cyg_selrecord( CYG_ADDRWORD info, struct CYG_SELINFO_TAG *sip )
{
//diag_printf("select.cxx cyg_selrecord Cyg_Thread\n");
    sip->si_info = info;
    sip->si_thread = (CYG_ADDRESS)Cyg_Thread::self();
}

// -------------------------------------------------------------------------
// cyg_selwakeup() is called when the client device matches the select
// criterion, and needs to wake up a selector.

void cyg_selwakeup( struct CYG_SELINFO_TAG *sip )
{
    // We don't actually use the si_info field of selinfo at present.
    // A potential use would be to select one of several selwait condition
    // variables to signal. However, that would only be necessary if we
    // end up having lots of threads in select.

    Cyg_Scheduler::lock();
 
    if( sip->si_thread != 0 )
    {
        // If the thread pointer is still present, this selection has
        // not been fired before. We just wake up all threads waiting,
        // regardless of whether they are waiting for this event or
        // not.  This avoids any race conditions, and is consistent
        // with the behaviour of the BSD kernel.
        
        sip->si_thread = 0;
        selwait.broadcast();
        selwake_count++;

    }

    Cyg_Scheduler::unlock();    
}

// -------------------------------------------------------------------------
// EOF select.cxx
